llama_index/indices/vector_store/sentence_window.py

```diff
@@ -0,0 +1,172 @@
+            )
+
+        all_nodes.extend(nodes)
+
+    return all_nodes
+
+
+class SentenceWindowVectorIndex(VectorStoreIndex):
+    """[BETA] Sentence Window Vector Index.
+
+    Note: This is a BETA feature, and is subject to change in future releases.
+    Some languages and text formats may not work well with this index.
+
+    Args:
+        use_async (bool): Whether to use asynchronous calls. Defaults to False.
+        show_progress (bool): Whether to show tqdm progress bars. Defaults to False.
+        store_nodes_override (bool): set to True to always store Node objects in index
+            store and document store even if vector store keeps text. Defaults to False
+    """
+
+    index_struct_cls = IndexDict
+    window_metadata_key = "window"
+
+    def __init__(
+        self,
+        nodes: Optional[Sequence[BaseNode]] = None,
+        index_struct: Optional[IndexDict] = None,
+        service_context: Optional[ServiceContext] = None,
+        storage_context: Optional[StorageContext] = None,
+        sentence_splitter: Optional[Callable[[str], List[str]]] = None,
+        window_size: int = 5,
+        use_async: bool = False,
+        store_nodes_override: bool = False,
+        show_progress: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize params."""
+        self._sentence_splitter = sentence_splitter or split_by_sentence_tokenizer()
+        self._window_size = window_size
+
+        super().__init__(
+            nodes=nodes,
+            index_struct=index_struct,
+            service_context=service_context,
+            storage_context=storage_context,
+            show_progress=show_progress,
+            store_nodes_override=store_nodes_override,
+            use_async=use_async,
+            **kwargs,
+        )
+
+    @classmethod
+    def from_documents(
+        cls: Type["SentenceWindowVectorIndex"],
+        documents: Sequence[Document],
+        storage_context: Optional[StorageContext] = None,
+        service_context: Optional[ServiceContext] = None,
+        show_progress: bool = False,
+        sentence_splitter: Optional[Callable[[str], List[str]]] = None,
+        window_size: int = 5,
+        **kwargs: Any,
+    ) -> "SentenceWindowVectorIndex":
+        """Create index from documents.
+
+        Args:
+            documents (Optional[Sequence[BaseDocument]]): List of documents to
+                build the index from.
+
+        """
+        storage_context = storage_context or StorageContext.from_defaults()
+        service_context = service_context or ServiceContext.from_defaults()
+        docstore = storage_context.docstore
+
+        sentence_splitter = sentence_splitter or split_by_sentence_tokenizer()
+
+        with service_context.callback_manager.as_trace("index_construction"):
+            for doc in documents:
+                docstore.set_document_hash(doc.get_doc_id(), doc.hash)
+            nodes = build_window_nodes_from_documents(
```

Comment:
this is technically a "parsing" step before any indexing happens.

is it possible to just have this as a node parser? and still reuse a vector store index? (but keep the retriever) 

i think i get the motivation for this abstraction but i'd want to decouple the idea of "creating parsed nodes" from the idea of "indexing" to create fewer inheritance dependnecies later on

Reply:
Yea I can refactor this part put into just a node parser. Slightly annoying for a user to setup (create node parser, create service context, pass it in), but let's see what it looks like

Reply:
Ended up refactoring this into a node parser + node post processor. Maybe a little more annoying to use, but also more general now